### PR TITLE
state: add machine storage for subordinates

### PR DESCRIPTION
**Pull request by axw**
When a subordinate unit is added, create its
machine storage when the principal unit is
assigned to a machine.

Fixes https://bugs.launchpad.net/juju/+bug/1595617

### CODE REVIEW

**Comment by axw** [comment]
**QA**

1. bootstrap lxd
2. deploy a principal charm
3. deploy a subordinate charm with filesystem storage
4. relate the two applications
5. when the principal unit is added, check that there is a storage instance created, and an associated filesystem

File: state/addmachine.go [hunk 1]
```diff
@@ -536,6 +536,31 @@ type machineStorageParams struct {
...
+	out := &machineStorageParams{}
+	out.volumes = append(lhs.volumes[:], rhs.volumes...)
+	out.filesystems = append(lhs.filesystems[:], rhs.filesystems...)
+	if lhs.volumeAttachments != nil || rhs.volumeAttachments != nil {
+		out.volumeAttachments = make(map[names.VolumeTag]VolumeAttachmentParams)
+		for k, v := range lhs.volumeAttachments {
+			out.volumeAttachments[k] = v
+		}
+		for k, v := range rhs.volumeAttachments {
+			out.volumeAttachments[k] = v
```
Comment (mjs) [inline state/addmachine.go]: I presume it's by design that the rhs wins if the same key appears in both sides?

File: state/unit.go [hunk 2]
```diff
@@ -1953,7 +2020,34 @@ func (u *Unit) findCleanMachineQuery(requireEmpty bool, cons *constraints.Value)
...
+	if err := u.st.run(buildTxn); err != nil {
+		return nil, errors.Trace(err)
+	}
+	u.doc.MachineId = m.doc.Id
+	m.doc.Clean = false
+	return m, nil
+}
+
+func (u *Unit) assignToCleanMaybeEmptyMachineOps(requireEmpty bool) (_ *Machine, _ []txn.Op, err error) {
+	failure := func(err error) (*Machine, []txn.Op, error) {
```
Comment (mjs) [inline state/unit.go]: nice pattern

File: state/addmachine.go (hunk 1 above)
Comment (axw) [inline state/addmachine.go]: Storage instances cannot share a volume/filesystem (attachment), so it's impossible.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**